### DISCUSSION TITLE
Memory Not Released After High-Load Operations

### DISCUSSION THREAD

**Discussion by EBazarov** [Kludex/uvicorn#2078] [Potential Issue] [answered]
### Summary:

We have developed an inference server using FastAPI with batching. However, we've encountered a significant issue related to memory management during high-load scenarios, in our case (described below) we are sending lists of base64-encoded images with a lot of simultaneous requests. Although we initially implemented this with FastAPI, further tests revealed that the issue also persists with a pure Starlette implementation, that’s why it’s submitted under the Starlette repository.
![memory profiler](https://github.com/Besedo/memory_issue/raw/main/test_gunicorn/memory_gunicorn_60m.png)
<p align=center> Memory consumption over time with Starlette + Gunicorn while waiting for 1 hour after the test ends.

### **Detailed Description:**

In our test scenario, the asynchronous server receives a request containing a list of base64 encoded images and simply returns “Hello World!”.

```python
from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.routing import Route

async def infer(request):
    data = await request.json()
    return JSONResponse({"message": "Hello World!"})

routes = [
    Route('/infer', infer, methods=['GET', 'POST'])
]

app = Starlette(debug=True, routes=routes)
```

**Current Behavior**:

- Memory consumption increases under high-load conditions, which is expected due to data backlog. However, the memory is **not released back to the system** even after the load decreases.

**Expected Behavior**:

- Memory consumption should increase under high-load conditions, but it should **revert to a level close to its initial state** once the high-load operations are complete.

### **Reproduction Steps:**

**[We have prepared a dedicated GitHub repository](https://github.com/Besedo/memory_issue/tree/main)** that showcases this issue in greater details. This includes **`mprof`** memory plots for each test, Dockerfiles and makefile for easy reproducibility. You can clone the repository and easily reproduce the issue following the README file with specific instructions.

TLDR what we did:

1. Initiate an asynchronous server using Starlette
2. Conduct a series of load tests (transmission of random length lists of base64 encoded images)
3. Monitor memory usage before, during, and after the load tests

### **What We've Tried:**

Some of the experiments can be found in the shared repository above, but overall what we tried:

- Switching to different versions of FastAPI, Starlette, and Python
- Utilizing various ASGI servers like hypercorn, uvicorn, and gunicorn
- Setting **`_default_thread_limiter`** in anyio to 1
- Waiting a few hours in idle, just to see if memory consumption will go down after the test load
- Forcing garbage collector request to remove not used objects
- We considered the possibility that this issue could stem from Python's own memory management, as there are documented cases where Python reserves memory from the OS. However, continuous monitoring of Resident Set Size (RSS) and Virtual Memory Size (VMS) at the Docker level showed that they stay almost at the same level

### Related issues

We found out that we are not the only ones having the same behaviour:

- **https://github.com/danielgatis/rembg/issues/289**
- **https://github.com/encode/uvicorn/issues/1564**
- **https://github.com/replicate/cog/issues/521**
- **https://github.com/tiangolo/fastapi/discussions/9145**
- **https://github.com/tiangolo/fastapi/discussions/9082**
- **https://github.com/tiangolo/fastapi/discussions/8612**
- **https://github.com/microsoft/planetary-computer-apis/pull/93**

**Comment by Kludex** [comment]
Does it persist with another ASGI web framework? 👀

**Reply by syrinecheriaa** [reply]
Nope we did not try other ASGI web framework, but we can do that. Do you have one/some that you recommend? (otherwise we will be trying Sanic).

**Reply by Kludex** [reply]
Sanic is good. Just use Gunicorn + Uvicorn, and not their server... 😅

**Reply by syrinecheriaa** [reply]
Hello @Kludex, thanks for your answer. 
We tested with Sanic and the same issue appears (we update the github repo with the results). 
Do you have any idea what is preventing the memory freeing? Is it something related to the http protocol? 
Is there something in the request context not being freed because the process is expecting the same amount of trafic again?

**Reply by Kludex** [reply]
I don't know yet... Now we know it's on the server side. It can either be `uvicorn` or `gunicorn`. Can you try removing `gunicorn` from the equation? Just use `uvicorn main:app` (and do NOT set the `--workers` flag on `uvicorn`, use just a single worker).

If the issue is solved, then it's `gunicorn`. If the issue is not solved, then it's `uvicorn`.

**Reply by syrinecheriaa** [reply]
The issue persists when using just uvicorn with 
`uvicorn main:app --host "$API_HOST" --port "$API_PORT"` (with both starlette and sanic).

**Reply by Kludex** [reply]
Uvicorn version?

**Reply by syrinecheriaa** [reply]
uvicorn  0.23.2

**Reply by Kludex** [reply]
I'm going to transfer this discussion to uvicorn.

**Reply by gi0baro** [reply]
The strange thing to me here is that Hypercorn seems to have the same issue.
So it seems:
- unrelated to the framework used
- unrelated to the server used

@Kludex I might run the same test with Granian and see if it makes any difference. If that succeed maybe there's something wrong with the protocol implementation `h11`..

**Reply by Kludex** [reply]
Ah! Can you try with httptools, @syrinecheriaa, please? Just install httptools, and uvicorn will use it.

Thanks @gi0baro 🙏

**Reply by gi0baro** [reply]
@Kludex from a rapid test it seems Granian doesn't suffer the same issue:

![memory_granian_py311](https://github.com/encode/uvicorn/assets/5019985/543891f4-141e-4878-9adb-c2b0e1a61354)

So I'm more prone to consider this an issue with the protocol implementation.

**Reply by syrinecheriaa** [reply]
Hello @Kludex, we tested uvicorn + httptools and the issue [persists](https://github.com/Besedo/memory_issue/tree/add_sanic/test_hypercorn).

**Reply by syrinecheriaa** [reply]
thank you @gi0baro, we will take a look at Granian.

**Reply by Kludex** [reply]
> Hello @Kludex, we tested uvicorn + httptools and the issue [persists](https://github.com/Besedo/memory_issue/tree/add_sanic/test_hypercorn).

You pasted a reference for `hypercorn`, not `httptools`...

**Reply by EBazarov** [reply]
@Kludex you are right here's the [right link](https://github.com/Besedo/memory_issue/tree/main/use_httptools)

**Reply by EBazarov** [reply]
> @Kludex from a rapid test it seems Granian doesn't suffer the same issue:
> 
> ![memory_granian_py311](https://user-images.githubusercontent.com/5019985/263665856-543891f4-141e-4878-9adb-c2b0e1a61354.png)
> 
> So I'm more prone to consider this an issue with the protocol implementation.

Hey @gi0baro, great result. Can you share your code with us, we just want to reproduce it and make sure we do things in the right way.

**Reply by gi0baro** [reply]
@EBazarov I just replaced `uvicorn` with `granian` from your starlette test with the command in the chart.
I can open up a PR in the repo if you want

**Reply by EBazarov** [reply]
understood, looks straightforward, let's us test it. Thank you

**Reply by Kludex** [reply]
It's extremely unlikely that there's a leak of such a magnitude with this simple app with the production setup. Try installing also `uvloop`, and see if still reproduces. If it reproduces, please improve the README, so I can run here. It's taking too long to print the chart.

**Reply by EBazarov** [reply]
@Kludex of course, not an issue, we will try to test it with `uvloop` also. I will say that it surprised me also, to see that, but still, we can see it in our local and production environments.
We will be happy to improve our README, let us know please where you think we should make changes, there's a makefile to reproduce the issue with two commands.

**Reply by gi0baro** [reply]
A quick correction, as I re-run the Granian tests. It seems the result I posted above was actually dependant on the fact the Rust workers died during the test.

Tweaking a bit the test itself so that Granian won't flood the Python event loop, shows results comparable with every other test:

![memory_granian](https://github.com/encode/uvicorn/assets/5019985/8d4134c1-6fd0-4c74-9fde-f9bcc631dc31)

So, at this point, given the number of frameworks and servers tested, I'm prone to give the culprit to Python itself..

**Reply by EBazarov** [reply]
Thank you for that, because we tested and still had the same issue, we were thinking of writing that after double-checking everything.

**Reply by syrinecheriaa** [reply]
@Kludex we updated the repo by adding a "quick test" that enables you to have a plot in less then one minute. Please refer to [this](https://github.com/Besedo/memory_issue/tree/main#reproduce-the-issue) part of the README to use it.

We also made a [test using `uvloop`](https://github.com/Besedo/memory_issue/tree/main/test_uvloop) and the issue persists.

**Comment by Kludex** [comment]
<img width="664" alt="Screenshot 2023-08-28 at 17 50 56" src="https://github.com/encode/uvicorn/assets/7353520/6f1d8359-7379-4769-91d8-456bbfb454d5">

This list is not useful. There was a leak in the past that was solved.

**Reply by EBazarov** [reply]
Agree that some of them were solved, but they contain a reference to things that we tried and it didn't worked for us, also we spotted that in some discussions people still have some troubles with memory consumption.

**Comment by Kludex** [comment]
Can you use the pipeline, and upload the artifact (image) from there?

**Reply by syrinecheriaa** [reply]
@Kludex you can pull the image from docker hub 
`docker pull syrinecher/test_memory_consumption:v0.0.1` (this image use starlette + uvicorn with the whole test (it should takes around 90 seconds to generate the plot))

**Comment by EBazarov** [comment]
Hi @Kludex,

Have you had a chance to check out the repository and reproduce the plots we've been working on? We're interested in digging deeper to understand the underlying issues better.

One approach we're considering is enhanced memory profiling. However, we'd really value your insights on this or any other suggestions you might have for uncovering the root cause.

Looking forward to your thoughts!

**Comment by ulan-yisaev** [comment]
Hi @EBazarov and team,

I've been closely following this thread as I'm currently facing similar memory issues with my app, especially during model inference. It's been quite a journey trying to pinpoint the exact cause.

While exploring solutions, I came across the memray profiler. It offers a detailed breakdown of memory consumption by different lines of code. For instance, to profile a FastAPI application using memray, one can use:
`memray run -m uvicorn app:app --workers 1`

I thought it might be worth sharing, just in case it could provide some insights or be of help in your investigations. Looking forward to any updates on this issue, and I truly appreciate all the efforts you and others are putting into this!

**Reply by hetpin** [reply]
We've added `memray` to the [repository/test_memray](https://github.com/Besedo/memory_issue/tree/test_memray/test_memray).
So far, memray's html report gives a great view of mem usage.
Thanks for your suggestion! @ulan-yisaev

**Reply by EBazarov** [reply]
Here's the plot:
 ![plot](https://github.com/Besedo/memory_issue/raw/main/test_memray/memory_memray.png)
cheers to @hetpin

**Reply by Kludex** [reply]
And the reason?

**Reply by EBazarov** [reply]
Hey @Kludex, we just shared the plots, and report that was been generated with `memray`, we still can't say why and explain the reasons that it has such behaviour. We suppose, but not sure, that it may be related to Python itself and memory management.

**Reply by suhjohn** [reply]
Is this reproduced in hypercorn for you @EBazarov ? For this exact problem potentially in #1993 I moved to hypercorn and I have not been seeing the leak. Also found a related discovery here: https://github.com/tiangolo/fastapi/issues/1624. Pretty sure this is not a Python problem and a Uvicorn problem

**Comment by james-ro-williams** [comment]
Hi everyone, any update on this? We're currently experiencing a similar issue.

**Reply by gi0baro** [reply, marked as answer]
Given everything we stated in the upper replies, I suspect there's no much to do about this. 
Given the number of frameworks and servers tested, it just seems the Python interpreter is the culprit here, as even if the heap memory gets released, the RSS remains high.

**Reply by Faolain** [reply]
Really appreciate the in-depth research done by everyone here but somehow this solution doesn't feel completely satisfactory/feels janky as a long term solution. Has this issue been raised in the CPython repo?  Update: Linked this issue https://github.com/python/cpython/issues/109534 with the thread above. Maybe related?

**Comment by Faolain** [comment]
@EBazarov apologies for the ping but would you be able to followup in the above thread https://github.com/python/cpython/issues/109534 as it seems with your tests you might have something already available to do so? Guido himself responded that the upstream team is hardpressed to debug this and if someone can produce a self contained example that demonstrates the issue it would go a long way to get it solved. I believe your repo goes quite a ways there. 🙏

**Comment by Kludex** [comment]
I'm locking this thread to avoid misguidance.

Since it reproduces on every ASGI server, then it's not a Uvicorn issue.
